NDTensors/src/blocksparse/blocksparse.jl

```diff
@@ -12,21 +14,29 @@ struct BlockSparse{ElT,VecT,N} <: TensorStorage{ElT}
   end
 end
 
-# TODO: Implement as `fieldtype(storagetype, :data)`.
-datatype(::Type{<:BlockSparse{<:Any,DataT}}) where {DataT} = DataT
-# TODO: Implement as `ndims(blockoffsetstype(storagetype))`.
-ndims(storagetype::Type{<:BlockSparse{<:Any,<:Any,N}}) where {N} = N
 # TODO: Implement as `fieldtype(storagetype, :blockoffsets)`.
 blockoffsetstype(storagetype::Type{<:BlockSparse}) = BlockOffsets{ndims(storagetype)}
 
-function set_datatype(storagetype::Type{<:BlockSparse}, datatype::Type{<:AbstractVector})
-  return BlockSparse{eltype(datatype),datatype,ndims(storagetype)}
+function TypeParameterAccessors.position(
+  ::Type{<:BlockSparse}, ::typeof(TypeParameterAccessors.parenttype)
+)
+  return Position(2)
+end
+TypeParameterAccessors.position(::Type{<:BlockSparse}, ::typeof(Base.ndims)) = Position(3)
+TypeParameterAccessors.parameter_name(::Type{<:BlockSparse}, ::Position{2}) = parenttype
+TypeParameterAccessors.parameter_name(::Type{<:BlockSparse}, ::Position{3}) = Base.ndims
+function TypeParameterAccessors.default_parameter(
+  type::Type{<:BlockSparse}, ::typeof(TypeParameterAccessors.parenttype)
+)
+  return Vector{default_parameter(type, eltype)}
 end
 
-function SetParameters.set_ndims(storagetype::Type{<:BlockSparse}, ndims::Int)
-  return BlockSparse{eltype(storagetype),datatype(storagetype),ndims}
+function TypeParameterAccessors.set_ndims(storagetype::Type{<:BlockSparse}, param::Int)
+  return set_parameter(storagetype, Base.ndims, param)
 end
 
+Base.ndims(type::Type{<:BlockSparse}) = parameter(type, Base.ndims)
```

Comment:
```suggestion
Base.ndims(type::Type{<:BlockSparse}) = parameter(type, ndims)
```

Reply:
Can't this just be a generic definition:
```julia
Base.ndims(type::Type) = parameter(type, ndims)
```
for any type?

Reply:
I don't think this is possible because `BlockSparse` is a `AbstractArray` so if we don't define `Base.ndims(type::Type{<:BlockSparse)` then it will use the `AbstractArray` version.

Reply:
Oh right, never mind. I was forgetting that would be type piracy, since Julia already defines `Base.ndims(::Type{<:AbstractArray})`.

Reply:
See my comment above, we may want to define a `TypeParameterAccessors.ndims` function.